Add Bloom Effect Smoothing hack

LunaMoo opened the pull request:
As a solution for #6192 ~ seems to work as planned in both D3D9 and OGL, but I'm better at breaking things than creating them, so please tell me, if the code can be simplified, I don't like repeating code, but not really sure how to do it otherwise.;]

LunaMoo committed:
Add Bloom Effect Smoothing hack

unknownbrackets commented on UI/GameSettingsScreen.cpp:
```diff
@@ -287,6 +287,11 @@ void GameSettingsScreen::CreateViews() {
 	depthRange->SetDisabledPtr(&g_Config.bSoftwareRendering);
 
+	static const char *bloomHackOptions[] = { "Off", "Safe", "Balanced", "Aggressive" };
+	PopupMultiChoice *bloomHack = graphicsSettings->Add(new PopupMultiChoice(&g_Config.iBloomHack, gs->T("Bloom Effect Smoothing"), bloomHackOptions, 0, ARRAY_SIZE(bloomHackOptions), gs, screenManager()));
```
I had no idea what bloom was until someone explained it to me in an issue here.  And I'm a generally knowledgeable doofus.

Maybe "Lower resolution for effects (reduces artifacts)"?

-[Unknown]

hrydgard replied:
It's a photographic term and perhaps a little obscure if you're not either a photographer or a game graphics programmer :) So yeah, you might be right, although a lot of PC gamers probably know what it is.

unknownbrackets commented on GPU/GLES/Framebuffer.cpp:
```diff
@@ -667,8 +667,44 @@ void FramebufferManager::ResizeFramebufFBO(VirtualFramebuffer *vfb, u16 w, u16 h
+			}
+			break;
+		}
+	}
```
Maybe this ought to be in some method on FramebufferCommon...

Anyway, I would structure it more like:

```c++
bool force1x = false;
switch (g_Config.iBloomHack) {
case 1:
   force1x = vfb->bufferWidth <= 128 || vfb->bufferHeight <= 64;
   break;
case 2:
   force1x = vfb->bufferWidth <= 256 || vfb->bufferHeight <= 128;
   break;
case 3:
   force1x = vfb->bufferWidth <= 480 || vfb->bufferHeight <= 272;
   break;
}

if (force1x && g_Config.iInternalResolution != 1) {
   vfb->renderWidth = vfb->bufferWidth;
   vfb->renderHeight = vfb->bufferHeight;
} else {
   vfb->renderWidth = vfb->bufferWidth * renderWidthFactor;
   vfb->renderHeight = vfb->bufferHeight * renderHeightFactor;
}
```

Either way, `g_Config.iInternalResolution==1` should have spaces.

Also, I wonder if `&&` for w/h would be better in some cases but doubt it matters...

-[Unknown]

LunaMoo committed:
Move the hack to FramebufferCommon.cpp and rename.

LunaMoo commented:
Moved it to FramebufferCommon and renamed, thanks for all the suggestions, looks much nicer now:)

 I don't think changing || to && would be good since from my testing it can squish it only along one dimension to make bloom like 256x272 and in other game it can even change it into small square like 64x64, maybe we can think about changing the presets after it turns out to be an issue, since it might as well not be and with affected games I tested it was just like that.:)

LunaMoo committed:
Oops, that's what I get for copy&paste:P

unknownbrackets commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -737,3 +737,29 @@ void FramebufferManagerCommon::NotifyBlockTransferAfter(u32 dstBasePtr, int dstS
+
+void FramebufferManagerCommon::SetRenderSize(VirtualFramebuffer *vfb) {
+	float renderWidthFactor = (float)PSP_CoreParameter().renderWidth / 480.0f;
+	float renderHeightFactor = (float)PSP_CoreParameter().renderHeight / 272.0f;
```
It doesn't really matter, but these lines can probably be removed now from `Framebuffer.cpp` and `FramebufferDX9.cpp` right?

-[Unknown]

LunaMoo committed:
Remove leftovers

LunaMoo commented:
Oh yeah, they were not used for anything else.:)

unknownbrackets commented:
Seems good to me.

-[Unknown]
